plugin/trino-pinot/src/main/java/io/trino/plugin/pinot/client/PinotGrpcDataFetcher.java

```diff
@@ -0,0 +1,289 @@
+        private static final CalciteSqlCompiler REQUEST_COMPILER = new CalciteSqlCompiler();
+
+        private final PinotHostMapper pinotHostMapper;
+        private final Map<HostAndPort, GrpcQueryClient> clientCache = new ConcurrentHashMap<>();
+        private final int grpcPort;
+        private final GrpcQueryClientFactory grpcQueryClientFactory;
+        private final Closer closer;
+
+        private PinotGrpcServerQueryClient(PinotHostMapper pinotHostMapper, PinotGrpcServerQueryClientConfig pinotGrpcServerQueryClientConfig, GrpcQueryClientFactory grpcQueryClientFactory, Closer closer)
+        {
+            this.pinotHostMapper = requireNonNull(pinotHostMapper, "pinotHostMapper is null");
+            this.grpcPort = requireNonNull(pinotGrpcServerQueryClientConfig, "pinotGrpcServerQueryClientConfig is null").getGrpcPort();
+            this.grpcQueryClientFactory = requireNonNull(grpcQueryClientFactory, "grpcQueryClientFactory is null");
+            this.closer = requireNonNull(closer, "closer is null");
+        }
+
+        public Iterator<PinotDataTableWithSize> queryPinot(ConnectorSession session, String query, String serverHost, List<String> segments)
+        {
+            HostAndPort mappedHostAndPort = pinotHostMapper.getServerGrpcHostAndPort(serverHost, grpcPort);
+            // GrpcQueryClient does not implement Closeable. The idle timeout is 30 minutes (grpc default).
+            GrpcQueryClient client = clientCache.computeIfAbsent(mappedHostAndPort, hostAndPort -> {
+                GrpcQueryClient queryClient = grpcQueryClientFactory.create(hostAndPort);
+                closer.register(queryClient::close);
+                return queryClient;
+            });
+            BrokerRequest brokerRequest = REQUEST_COMPILER.compileToBrokerRequest(query);
+            GrpcRequestBuilder requestBuilder = new GrpcRequestBuilder()
+                    .setSql(query)
+                    .setSegments(segments)
+                    .setEnableStreaming(true)
+                    .setBrokerRequest(brokerRequest);
+            return new ResponseIterator(client.submit(requestBuilder.build()));
+        }
+
+        public static class ResponseIterator
+                extends AbstractIterator<PinotDataTableWithSize>
+        {
+            private final Iterator<Server.ServerResponse> responseIterator;
+
+            public ResponseIterator(Iterator<Server.ServerResponse> responseIterator)
+            {
+                this.responseIterator = requireNonNull(responseIterator, "responseIterator is null");
+            }
+
+            @Override
+            protected PinotDataTableWithSize computeNext()
+            {
+                if (!responseIterator.hasNext()) {
+                    return endOfData();
+                }
+                Server.ServerResponse response = responseIterator.next();
+                String responseType = response.getMetadataMap().get(CommonConstants.Query.Response.MetadataKeys.RESPONSE_TYPE);
```

Comment:
Import MetadataKeys and Response type so that these checks are a little easier to read:

```java
String responseType = response.getMetadataMap().get(MetadataKeys.RESPONSE_TYPE);
if (responseType.equals(ResponseType.METADATA)) {
   ...
```